### PR TITLE
N2n 194

### PR DESCRIPTION
**Pull request by zippy**
Implementing this ticket, led us down the path of realizing that what we were calling the nucleii (the different vm/code execution environments) should really be called the ribosomes, i.e. machinery for interpreting the code held in the zomes.  The nucleus then becomes singular (as it should be) and is an abstraction for holding the various elements that correspond to the "app".  

So in this ticket you'll see all the stuff that was call Nucleus renamed to Ribosome, and a new nucleus abstraction that holds the messaging protocols that need to all into the nucleus (i.e. validation & application) kind of like the DHT abstraction holds the dht and gossip protocols.  

This also allowed creating zome.go file and moving some of the zome related stuff that was in holochain.go into its own more logically coherent location.

### CODE REVIEW

File: zygosome.go [hunk 1]
```diff
@@ -549,6 +582,39 @@ func NewZygoNucleus(h *Holochain, code string) (n Nucleus, err error) {
...
+
+			var r interface{}
+			r, err = a.Do(h)
+			var resp zygo.Sexp
+			if err == nil {
+				resp = &zygo.SexpStr{S: r.(string)}
+			}
+			return makeResult(env, resp, err)
+		})
+
```
Comment (lucksus) [inline zygosome.go]: Hm.. There is quite some duplicated code between the ribosomes. The send function in the JS ribosome looks almost the same. And actually that is the case for most ribosome functions. Would like to see some way to reduce the redundancy here in general, but that is maybe out of the scope of this PR.

File: zome.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	return
+}
+
+func (zome *Zome) MakeRibosome(h *Holochain) (r Ribosome, err error) {
+	//check to see if we have a cached version of the code, otherwise read from disk
+	if zome.code == "" {
+		zpath := h.ZomePath(zome)
+		var code []byte
+
+		code, err = readFile(zpath, zome.Code)
```
Comment (zippy) [inline zome.go]: Yes, but I would like to have this change happen as part of #248.

File: nucleus.go [hunk 3]
```diff
@@ -1,105 +1,67 @@
...
-			available = append(available, k)
+// AppReceiver handles messages on the application protocol
+func AppReceiver(h *Holochain, msg *Message) (response interface{}, err error) {
+	switch msg.Type {
+	case APP_MESSAGE:
+		h.config.Loggers.App.Logf("got app message: %v", msg)
+		switch t := msg.Body.(type) {
+		case AppMsg:
+			var r Ribosome
+			r, _, err = h.MakeRibosome(t.ZomeType)
```
Comment (zippy) [inline nucleus.go]: Absolutely, will do.

File: zygosome.go (hunk 1 above)
Comment (zippy) [inline zygosome.go]: Yep, this needs a refactor.  Added #250 and also see #157

File: nucleus.go (hunk 3 above)
Comment (lucksus) [inline nucleus.go]: Or I will now :)

**Pull request merged**